app/grandchallenge/cases/models.py

```diff
@@ -948,3 +983,85 @@ def start_dicom_import_job(self):
             )
+
+    def get_job_output_manifest(self, *, event):
+        output_uri = event["outputS3Uri"]
+        parsed = urlparse(output_uri)
+        bucket = parsed.netloc
+        key = parsed.path.lstrip("/") + "job-output-manifest.json"
+
+        # Try to get the manifest.
+        try:
+            obj = self._s3_client.get_object(Bucket=bucket, Key=key)
+        except self._s3_client.exceptions.NoSuchKey as e:
+            raise RetryStep("Manifest not (yet) found for job output") from e
+
+        return json.load(obj["Body"])["jobSummary"]
+
+    def get_job_output_failure_log(self, *, manifest):
+        output_uri = manifest["failureOutputS3Uri"]
+        parsed = urlparse(output_uri)
+        bucket = parsed.netloc
+        key = parsed.path.lstrip("/") + "failure.ndjson"
+
+        obj = self._s3_client.get_object(Bucket=bucket, Key=key)
+
+        return [json.loads(line) for line in obj["Body"]]
+
+    def handle_event(self, *, event):
+        try:
+            job_status = event["jobStatus"]
+            if job_status == "COMPLETED":
+                self.handle_completed_job(event=event)
+            elif job_status == "FAILED":
+                self.handle_failed_job(event=event)
+            else:
+                raise ValueError("Invalid job status")
+        except Exception as e:
+            self._mark_failed(
+                error_message="An unexpected error occurred", exc=e
+            )
+        else:
+            self.status = self.DICOMImageSetUploadStatusChoices.COMPLETED
+            self.save()
+
+    def handle_completed_job(self, *, event):
+        manifest = self.get_job_output_manifest(event=event)
+        if manifest["numberOfGeneratedImageSets"] > 1:
+            self.cleanup_image_sets(manifest=manifest)
+            raise DICOMImportJobValidationError(
+                "Multiple image sets created. Expected only one."
+            )
+        image_set = manifest["imageSetsSummary"][0]
+        if not image_set["isPrimary"]:
+            self.cleanup_image_sets(manifest=manifest)
+            raise DICOMImportJobValidationError(
+                "New instance is not primary: "
+                "metadata conflicts with already existing instance."
+            )
+        if not image_set["imageSetVersion"] == 1:
+            self.revert_image_set_to_initial_version(image_set)
+            raise DICOMImportJobValidationError(
+                "Instance already exists. This should never happen!"
+            )
+        self.convert_image_set_to_internal(image_set)
+
+    def handle_failed_job(self, *, event):
+        manifest = self.get_job_output_manifest(event=event)
+        self.cleanup_image_sets(manifest=manifest)
+        failure_log = self.get_job_output_failure_log(manifest=manifest)
+        job_id = manifest["jobId"]
+        raise DICOMImportJobFailedError(
+            message=f"Import job {job_id} failed", message_details=failure_log
```

Comment:
What is the intention of message_details? If it is for the user they should not just be sent logs.

Reply:
Okay - then it might be worth storing that in a separate field on the model, `internal_failure_log`, etc.